### PR TITLE
KYLIN-4683 Fail to consume kafka when partition number get larger

### PR DESCRIPTION
**Pull request by tianhui5**
## Proposed changes

JIRA：https://issues.apache.org/jira/browse/KYLIN-4683

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [x] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [ ] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

### CODE REVIEW

**Comment by tianhui5** [comment]
I think there is some network failure of CI

**Comment by hit-lacus** [comment]
Thank you for your contributions.

**Comment by hit-lacus** [comment]
Let's do a test to check what should we do when we need to scale up the topic partitions.

Give we have a topic `useraction_xxyu`, it has ten partitions.
image
image

```sh
[root@cdh-client xiaoxiang.yu]# kafka-topics  --describe --topic useraction_xxyu --zookeeper cdh-master:2181

...
...

20/08/09 22:19:36 INFO zookeeper.ZooKeeperClient: [ZooKeeperClient] Connected.
Topic:useraction_xxyu	PartitionCount:10	ReplicationFactor:1	Configs:
	Topic: useraction_xxyu	Partition: 0	Leader: 116	Replicas: 116	Isr: 116
	Topic: useraction_xxyu	Partition: 1	Leader: 114	Replicas: 114	Isr: 114
	Topic: useraction_xxyu	Partition: 2	Leader: 115	Replicas: 115	Isr: 115
	Topic: useraction_xxyu	Partition: 3	Leader: 116	Replicas: 116	Isr: 116
	Topic: useraction_xxyu	Partition: 4	Leader: 114	Replicas: 114	Isr: 114
	Topic: useraction_xxyu	Partition: 5	Leader: 115	Replicas: 115	Isr: 115
	Topic: useraction_xxyu	Partition: 6	Leader: 116	Replicas: 116	Isr: 116
	Topic: useraction_xxyu	Partition: 7	Leader: 114	Replicas: 114	Isr: 114
	Topic: useraction_xxyu	Partition: 8	Leader: 115	Replicas: 115	Isr: 115
	Topic: useraction_xxyu	Partition: 9	Leader: 116	Replicas: 116	Isr: 116
```

**Comment by hit-lacus** [comment]
Before we scale up the topic partition , we should doing the following steps:
1. Disable the cube, thus all consumption task will be cacelled.

**Comment by hit-lacus** [comment]
1. I will kill the producer.

image

2. check latest offset via Python

image

3. Resize partition count.

```sh
[root@cdh-client xiaoxiang.yu]# kafka-topics --alter --topic useraction_xxyu --zookeeper cdh-master:2181 --partitions 16
20/08/09 23:13:33 INFO utils.Log4jControllerRegistration$: Registered kafka:type=kafka.Log4jController MBean
20/08/09 23:13:34 INFO zookeeper.ZooKeeperClient: [ZooKeeperClient] Initializing a new session to cdh-master:2181.
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:zookeeper.version=3.4.5-cdh5.14.2--1, built on 03/27/2018 20:39 GMT
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:host.name=cdh-client
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:java.version=1.8.0_171
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:java.vendor=Oracle Corporation
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:java.home=/usr/java/jdk1.8.0_171/jre
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:java.library.path=/usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:java.io.tmpdir=/tmp
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:java.compiler=<NA>
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:os.name=Linux
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:os.arch=amd64
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:os.version=2.6.32-754.17.1.el6.x86_64
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:user.name=root
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:user.home=/root
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Client environment:user.dir=/root/xiaoxiang.yu
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Initiating client connection, connectString=cdh-master:2181 sessionTimeout=30000 watcher=kafka.zookeeper.ZooKeeperClient$ZooKeeperClientWatcher$@6302bbb1
20/08/09 23:13:34 INFO zookeeper.ZooKeeperClient: [ZooKeeperClient] Waiting until connected.
20/08/09 23:13:34 INFO zookeeper.ClientCnxn: Opening socket connection to server cdh-master/10.1.3.90:2181. Will not attempt to authenticate using SASL (unknown error)
20/08/09 23:13:34 INFO zookeeper.ClientCnxn: Socket connection established, initiating session, client: /10.1.3.91:54122, server: cdh-master/10.1.3.90:2181
20/08/09 23:13:34 INFO zookeeper.ClientCnxn: Session establishment complete on server cdh-master/10.1.3.90:2181, sessionid = 0x172bdedb56fcc33, negotiated timeout = 30000
20/08/09 23:13:34 INFO zookeeper.ZooKeeperClient: [ZooKeeperClient] Connected.
WARNING: If partitions are increased for a topic that has a key, the partition logic or ordering of the messages will be affected
20/08/09 23:13:34 INFO zk.AdminZkClient: Creating 6 partitions for 'useraction_xxyu' with the following replica assignment: Map(11 -> ArrayBuffer(115), 14 -> ArrayBuffer(115), 13 -> ArrayBuffer(114), 10 -> ArrayBuffer(114), 12 -> ArrayBuffer(116), 15 -> ArrayBuffer(116)).
20/08/09 23:13:34 INFO zk.AdminZkClient: Topic update Map(useraction_xxyu-8 -> Vector(115), useraction_xxyu-6 -> Vector(116), useraction_xxyu-0 -> Vector(116), useraction_xxyu-13 -> ArrayBuffer(114), useraction_xxyu-1 -> Vector(114), useraction_xxyu-9 -> Vector(116), useraction_xxyu-15 -> ArrayBuffer(116), useraction_xxyu-11 -> ArrayBuffer(115), useraction_xxyu-5 -> Vector(115), useraction_xxyu-2 -> Vector(115), useraction_xxyu-10 -> ArrayBuffer(114), useraction_xxyu-14 -> ArrayBuffer(115), useraction_xxyu-7 -> Vector(114), useraction_xxyu-12 -> ArrayBuffer(116), useraction_xxyu-4 -> Vector(114), useraction_xxyu-3 -> Vector(116))
Adding partitions succeeded!
20/08/09 23:13:34 INFO zookeeper.ZooKeeperClient: [ZooKeeperClient] Closing.
20/08/09 23:13:34 INFO zookeeper.ZooKeeper: Session: 0x172bdedb56fcc33 closed
20/08/09 23:13:34 INFO zookeeper.ClientCnxn: EventThread shut down
20/08/09 23:13:34 INFO zookeeper.ZooKeeperClient: [ZooKeeperClient] Closed.
```

**Comment by hit-lacus** [comment]
1. Go to web page and re-enable Cube, you will see some failure, please don't be in a panic.
```sh
2020-08-09 23:23:10,437 INFO  [qtp193625667-35] controller.AdminController:58 : receive assign request:AssignRequest{cubeName='UserAnalysisCube', partitions=[Partition [partitionId=0, partitionInfo=null], Partition [partitionId=5, partitionInfo=null], Partition [partitionId=10, partitionInfo=null], Partition [partitionId=15, partitionInfo=null], Partition [partitionId=13, partitionInfo=null], Partition [partitionId=2, partitionInfo=null], Partition [partitionId=8, partitionInfo=null], Partition [partitionId=12, partitionInfo=null], Partition [partitionId=14, partitionInfo=null], Partition [partitionId=9, partitionInfo=null], Partition [partitionId=11, partitionInfo=null], Partition [partitionId=1, partitionInfo=null], Partition [partitionId=4, partitionInfo=null], Partition [partitionId=6, partitionInfo=null], Partition [partitionId=7, partitionInfo=null], Partition [partitionId=3, partitionInfo=null]], startConsumers=true}
2020-08-09 23:23:10,437 INFO  [qtp193625667-35] server.StreamingServer:351 : create and start new consumer for cube:UserAnalysisCube
2020-08-09 23:23:10,468 INFO  [qtp193625667-35] kafka.KafkaSource:141 : Create StreamingConnector for Cube UserAnalysisCube, assignedPartitions [Partition [partitionId=0, partitionInfo=null], Partition [partitionId=5, partitionInfo=null], Partition [partitionId=10, partitionInfo=null], Partition [partitionId=15, partitionInfo=null], Partition [partitionId=13, partitionInfo=null], Partition [partitionId=2, partitionInfo=null], Partition [partitionId=8, partitionInfo=null], Partition [partitionId=12, partitionInfo=null], Partition [partitionId=14, partitionInfo=null], Partition [partitionId=9, partitionInfo=null], Partition [partitionId=11, partitionInfo=null], Partition [partitionId=1, partitionInfo=null], Partition [partitionId=4, partitionInfo=null], Partition [partitionId=6, partitionInfo=null], Partition [partitionId=7, partitionInfo=null], Partition [partitionId=3, partitionInfo=null]], startProtocol null
2020-08-09 23:23:10,470 INFO  [qtp193625667-35] kafka.TimedJsonStreamParser:98 : Using parser field mapping by {content_list=content_list, act_type=act_type, event_ts_2=event_ts_2, event_ts=event_ts, devide_type=user_detail.devide_type, location_city=user_detail.location.city, network_type=user_detail.network_type, video_id=video_id, event_date_2=event_date_2, str_minute=str_minute, video_type=video_type, play_times=play_times, pageview_id=pageview_id, active_minutes=active_minutes, device_brand=device_brand, str_minute_second=str_minute_second, play_duration=play_duration, event_date=event_date, page_id=page_id, str_second=str_second, uid=uid}
2020-08-09 23:23:10,470 INFO  [qtp193625667-35] kafka.TimedJsonStreamParser:120 : TimedJsonStreamParser with formatTs false tsColName event_date
2020-08-09 23:23:10,470 INFO  [qtp193625667-35] consumer.ConsumerConfig:279 : ConsumerConfig values:
	auto.commit.interval.ms = 5000
	auto.offset.reset = latest
	bootstrap.servers = [cdh-master:9092, cdh-worker-1:9092, cdh-worker-2:9092]
	check.crcs = true
	client.dns.lookup = default
	client.id =
	connections.max.idle.ms = 540000
	default.api.timeout.ms = 60000
	enable.auto.commit = false
	exclude.internal.topics = true
	fetch.max.bytes = 52428800
	fetch.max.wait.ms = 500
	fetch.min.bytes = 1
	group.id =
	heartbeat.interval.ms = 3000
	interceptor.classes = []
	internal.leave.group.on.close = true
	isolation.level = read_uncommitted
	key.deserializer = class org.apache.kafka.common.serialization.ByteArrayDeserializer
	max.partition.fetch.bytes = 1048576
	max.poll.interval.ms = 300000
	max.poll.records = 500
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	partition.assignment.strategy = [class org.apache.kafka.clients.consumer.RangeAssignor]
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 30000
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = null
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.mechanism = GSSAPI
	security.protocol = PLAINTEXT
	send.buffer.bytes = 131072
	session.timeout.ms = 20000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = https
	ssl.key.password = null
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.location = null
	ssl.keystore.password = null
	ssl.keystore.type = JKS
	ssl.protocol = TLS
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.location = null
	ssl.truststore.password = null
	ssl.truststore.type = JKS
	value.deserializer = class org.apache.kafka.common.serialization.ByteArrayDeserializer

2020-08-09 23:23:10,473 INFO  [qtp193625667-35] utils.AppInfoParser:109 : Kafka version : 2.1.1
2020-08-09 23:23:10,473 INFO  [qtp193625667-35] utils.AppInfoParser:110 : Kafka commitId : 21234bee31165527
2020-08-09 23:23:10,475 INFO  [qtp193625667-35] kafka.KafkaSource:192 : localConsumeStats from local checkpoint null, remoteConsumeStats from remote checkpoint {"0":363171,"1":363198,"2":363249,"3":363171,"4":363199,"5":363250,"6":363170,"7":363196,"8":363250,"9":363170}
2020-08-09 23:23:10,475 INFO  [qtp193625667-35] kafka.KafkaSource:239 : start kafka connector from specified position:KafkaPosition{partitionOffsetMap={0=363171, 1=363198, 2=363249, 3=363171, 4=363199, 5=363250, 6=363170, 7=363196, 8=363250, 9=363170}}
2020-08-09 23:23:10,476 ERROR [qtp193625667-35] server.StreamingServer:357 : consumer start fail for cube:UserAnalysisCube
java.lang.NullPointerException
	at org.apache.kylin.stream.source.kafka.consumer.KafkaConnector.open(KafkaConnector.java:86)
	at org.apache.kylin.stream.core.consumer.StreamingConsumerChannel.start(StreamingConsumerChannel.java:85)
	at org.apache.kylin.stream.server.StreamingServer.startConsumer(StreamingServer.java:355)
	at org.apache.kylin.stream.server.StreamingServer.startConsumers(StreamingServer.java:327)
	at org.apache.kylin.stream.server.rest.controller.AdminController.assign(AdminController.java:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:133)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:97)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:849)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:760)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:967)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:901)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:872)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:755)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:547)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1297)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:485)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1212)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:221)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:270)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.lang.Thread.run(Thread.java:748)
```
2. Let try doing a reassignment.

image

image

image

File: stream-source-kafka/src/main/java/org/apache/kylin/stream/source/kafka/consumer/KafkaConnector.java [hunk 1]
```diff
@@ -81,9 +81,15 @@ public void open() {
...
+            List<TopicPartition> newTopicPartitions = Lists.newArrayList();
             for (TopicPartition topicPartition : topicPartitions) {
                 Long offset = partitionOffsets.get(topicPartition.partition());
-                kafkaConsumer.seek(topicPartition, offset);
+                if (offset != null) {
+                    kafkaConsumer.seek(topicPartition, offset);
+                } else {
+                    newTopicPartitions.add(topicPartition);
+                }
+                kafkaConsumer.seekToBeginning(newTopicPartitions);
```
Comment (hit-lacus) [inline stream-source-kafka/src/main/java/org/apache/kylin/stream/source/kafka/consumer/KafkaConnector.java:92]: I think `kafkaConsumer.seekToBeginning(newTopicPartitions);` should not in **for-loop** .

Comment (tianhui5) [reply stream-source-kafka/src/main/java/org/apache/kylin/stream/source/kafka/consumer/KafkaConnector.java:92]: My mistake, thanks.

**Comment by tianhui5** [comment]
> Thank you for you effort that you spend on this issue, here is my suggestion, what do you think?
> If `offset` is null, please just set it to zero.

I wonder if there is any difference between setting offset to zero and seekToBeginning.

**Comment by hit-lacus** [comment]
> > Thank you for you effort that you spend on this issue, here is my suggestion, what do you think?
> > If `offset` is null, please just set it to zero.
> 
> I wonder if there is any difference between setting offset to zero and seekToBeginning.

Please correct me if I was wrong.
Let take the above test by me as an example, after scale up, it should continue its consumptation from this position:

```json
{
    "0":363171,
    "1":363198,
    "2":363249,
    "3":363171,
    "4":363199,
    "5":363250,
    "6":363170,
    "7":363196,
    "8":363250,
    "9":363170,
    "10":0,
    "11":0,
    "12":0,
    "13":0,
    "14":0,
    "15":0
}
```

not this position:

```json
{
    "0":0,
    "1":0,
    "2":0,
    "3":0,
    "4":0,
    "5":0,
    "6":0,
    "7":0,
    "8":0,
    "9":0,
    "10":0,
    "11":0,
    "12":0,
    "13":0,
    "14":0,
    "15":0
}
```

So, if you are using `seekToBeginning` API, can you make sure the partition 0 to partition 9 consume at right position?

**Comment by tianhui5** [comment]
> > > Thank you for you effort that you spend on this issue, here is my suggestion, what do you think?
> > > If `offset` is null, please just set it to zero.
> > 
> > 
> > I wonder if there is any difference between setting offset to zero and seekToBeginning.
> 
> Please correct me if I was wrong.
> Let take the above test by me as an example, after scale up, it should continue its consumptation from this position:
> 
> ```json
> {
>     "0":363171,
>     "1":363198,
>     "2":363249,
>     "3":363171,
>     "4":363199,
>     "5":363250,
>     "6":363170,
>     "7":363196,
>     "8":363250,
>     "9":363170,
>     "10":0,
>     "11":0,
>     "12":0,
>     "13":0,
>     "14":0,
>     "15":0
> }
> ```
> 
> not this position:
> 
> ```json
> {
>     "0":0,
>     "1":0,
>     "2":0,
>     "3":0,
>     "4":0,
>     "5":0,
>     "6":0,
>     "7":0,
>     "8":0,
>     "9":0,
>     "10":0,
>     "11":0,
>     "12":0,
>     "13":0,
>     "14":0,
>     "15":0
> }
> ```
> 
> So, if you are using `seekToBeginning` API, can you make sure the partition 0 to partition 9 consume at right position?

Actually, I just add new partitions to list which seekToBeging. So the old partitions' offset won't be effected. Here is my test when I scale kafka from 7 partitions to 10 partitions.

2020-08-21 14:18:02,634 INFO  [qtp310350177-45] kafka.KafkaSource:192 : localConsumeStats from local checkpoint null, remoteConsumeStats from remote checkpoint {"0":1369,"1":1412,"2":1500,"3":1395,"4":1430,"5":0,"6":0}

2020-08-21 14:24:33,038 INFO  [cost_test_channel] storage.StreamingSegmentManager:645 : Print check point for cube cost_test ,CheckPoint{sourceConsumePosition='{"0":1408,"4":1476,"9":0}', persistedIndexes={1597978800000=1, 1597986000000=1, 1597982400000=1, 1597975200000=1, 1597989600000=3}, longLatencyInfo=LongLatencyInfo{longLatencyEventCnts={}, totalLongLatencyEventCnt=0}, segmentSourceStartPosition={1597978800000={"0":1376,"4":1442}, 1597986000000={"0":1398,"4":1457}, 1597982400000={"0":1392,"4":1454}, 1597975200000={"0":1369,"4":1430}, 1597989600000={"0":1408,"4":1467,"9":0}}, checkPointTime=1597991073038, totalCount=88, checkPointCount=1}

**Review by hit-lacus — APPROVED** [review]

**Comment by hit-lacus** [comment]
LGTM

**Pull request merged**